extensions/amp-wistia-player/0.1/amp-wistia-player.js

```diff
@@ -88,25 +119,88 @@ class AmpWistiaPlayer extends AMP.BaseElement {
       this.iframe_ = null;
     }
 
+    if (this.unlistenMessage_) {
+      this.unlistenMessage_();
+    }
+
+    this.playerReadyPromise_ = new Promise(resolve => {
+      this.playerReadyResolver_ = resolve;
+    });
+
     return true;
   }
 
+  /** @override */
+  viewportCallback(visible) {
+    this.element.dispatchCustomEvent(VideoEvents.VISIBILITY, {visible});
+  }
+
   /** @override */
   pauseCallback() {
     if (this.iframe_) {
       this.pause();
     }
   }
 
+  /** @private */
+  handleWistiaMessages_(event) {
+    if (event.origin !== 'https://fast.wistia.net' ||
+      event.source != this.iframe_.contentWindow) {
+      return;
+    }
+
+    if (!getData(event) || !(isObject(getData(event))
+        || startsWith(/** @type {string} */ (getData(event)), '{'))) {
+      return; // Doesn't look like JSON.
+    }
+    /** @const {?JsonObject} */
+    const data = /** @type {?JsonObject} */ (isObject(getData(event))
+      ? getData(event)
+      : tryParseJson(getData(event)));
+
+    if (data === undefined) {
+      return; // We only process valid JSON.
+    }
+
+    if (data['method'] == 'statechange') {
+      state = (data['args'] ? data['args']['_state'] : undefined)
+      if (state === 'playing') {
+        this.element.dispatchCustomEvent(VideoEvents.PLAYING);
+      } else if (state === 'paused') {
+        this.element.dispatchCustomEvent(VideoEvents.PAUSE);
+      } else if (state === 'ended') {
+        this.element.dispatchCustomEvent(VideoEvents.PAUSE);
+        this.element.dispatchCustomEvent(VideoEvents.ENDED);
+      }
+    } else if (data['method'] == 'mutechange') {
```

Comment:
This is kinda semantics, but to me, `mutechange` is a player event. I think it confuses me because our iframe API passes an argument called `method` which actually causes a method to be called.

Reply:
okay I suppose I could make a second postMessage thing inside `_inside_iframe.coffee` that passes messages in a slightly different format rather than re-using the pre-existing one.

Reply:
Which pre-existing format are you referring to--the one in amp or the one in our player? Just a bit confused. Any relevant code links?

Reply:
https://github.com/wistia/wistia/blob/anders/add-amp-support-to-iframe-embed/player/modern/src/external/E-v1/_inside_iframe.coffee#L102

Reply:
Ah, I see what you're looking at now. That's that `method` code I was talking about too. It would literally call the `updateProperties` method on the iframe API object. `_isMuted` gets updated like so: https://github.com/wistia/wistia/blob/c8a7e1ce3ceaad43d6854dd4f2691c3c60230ac5/player/modern/src/external/E-v1/_inside_iframe.coffee#L130-L131

If you want to leverage the existing events, you can look at _trigger, which is what we use to post up most events from the player: https://github.com/wistia/wistia/blob/c8a7e1ce3ceaad43d6854dd4f2691c3c60230ac5/player/modern/src/external/E-v1/_inside_iframe.coffee#L169.

Reply:
To follow up, I think that, as it stands, this code would cause an error if E-v1 was on the same page as the AMP script because the `mutechange` method is not defined on the iframe api object: https://github.com/wistia/wistia/blob/c8a7e1ce3ceaad43d6854dd4f2691c3c60230ac5/player/modern/src/external/E-v1/_iframe_api.coffee#L659